File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
    _, maxDiameter := dfs_helper(root)
    return maxDiameter
}


func dfs_helper(root *TreeNode) (int,int){
    if root == nil{
        return 0,0
    }

    Ldepth, Ldiameter := dfs_helper(root.Left)
    Rdepth, Rdiameter := dfs_helper(root.Right)
    depth := 1 + max(Ldepth, Rdepth)

    diameter := Ldepth + Rdepth
    maxDiameter := max(diameter, max(Ldiameter, Rdiameter))


    return depth, maxDiameter
}


func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: LeetCode/Easy/0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
    int maxD = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        maxDepth(root);
        return maxD;
    }
    int maxDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int lD = maxDepth(root->left);
        int rD = maxDepth(root->right);
        maxD = max(maxD, lD + rD);
        return (1+max(lD, rD));
    }
};